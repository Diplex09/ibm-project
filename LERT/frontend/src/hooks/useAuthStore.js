import axios from "axios";
import { useSelector, useDispatch } from "react-redux";
import {
    authLogin,
    authLogout,
    authCheckingFinish,
} from "../reducers/authSlice";

const baseUrl = "https://lert-api.mybluemix.net";
// const baseUrl = "http://localhost:5000";

export const useAuthStore = () => {
    const { checking, uid, name, rol, rolName } = useSelector(
        (state) => state.auth
    );
    const dispatch = useDispatch();

    const startLogin = async (email, password) => {
        await axios
            .post(`${baseUrl}/login`, {
                email,
                password,
            })
            .then((resp) => {
                const { user } = resp.data;
                dispatch(
                    authLogin({
                        uid: user.uid,
                        name: user.fullName,
                        rol: user.rol,
                        rolName: user.rolName,
                    })
                );
            })
            .catch((error) => {
                if (error.response) {
                    // console.log(error.response.data);
                    checkingFinish();
                }
            });
    };

    const startLogout = async () => {
        await axios
            .get(`${baseUrl}/logout`)
            .then((resp) => {
                dispatch(authLogout());
            })
            .catch((error) => {
                console.log(error);
            });
    };

    const startChecking = async () => {
        await axios
            .get(`${baseUrl}/check`)
            .then((resp) => {
                console.log(resp);
                const { user } = resp.data;
                dispatch(
                    authLogin({
                        uid: user.uid,
                        name: user.fullName,
                        rol: user.rol,
                        rolName: user.rolName,
                    })
                );
            })
            .catch((error) => {
                if (window.location.pathname !== "/login") {
                    startLogout();
                }
                checkingFinish();
            });
    };

    const checkingFinish = () => {
        dispatch(authCheckingFinish());
    };

    return {
        // Properties
        checking,
        uid,
        name,
        rol,
        rolName,

        // Methods
        startLogin,
        startLogout,
        startChecking,
        checkingFinish,
    };
};
