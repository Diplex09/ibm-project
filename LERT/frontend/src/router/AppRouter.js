import { useDispatch, useSelector } from 'react-redux';
import { Routes, Route, BrowserRouter } from 'react-router-dom';

import { PublicRoute } from './PublicRoute';
import { PrivateRoute } from './PrivateRoute';
import Login from '../components/Login';
import MainView from '../MainView';
import { useEffect } from 'react';
import { startChecking } from '../actions/auth';

export const AppRouter = () => {
    const dispatch = useDispatch();
    const { checking, uid } = useSelector((state) => state.auth);

    useEffect(() => {
        dispatch(startChecking());
    }, []);

    if (checking) {
        return <h5>Loading...</h5>;
    }

    return (
        <BrowserRouter>
            <Routes>
                <Route
                    path="/login"
                    element={
                        <PublicRoute isAuthenticated={!!uid}>
                            <Login />
                        </PublicRoute>
                    }
                />
                <Route
                    path="/*"
                    element={
                        <PrivateRoute isAuthenticated={!!uid}>
                            <MainView />
                        </PrivateRoute>
                    }
                />
            </Routes>
        </BrowserRouter>
    );
};
