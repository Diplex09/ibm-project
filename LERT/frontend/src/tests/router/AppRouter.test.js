import { render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { useAuthStore } from "../../hooks/useAuthStore";
import { AppRouter } from "../../router/AppRouter";
import { testUserInfo } from "../fixtures/testUser";

jest.mock("../../hooks/useAuthStore");

describe("Tests in AppRouter", () => {
    const mockStartChecking = jest.fn();

    beforeEach(() => jest.clearAllMocks());

    test("should show Loading screen and call startChecking", () => {
        useAuthStore.mockReturnValue({
            checking: true,
            startChecking: mockStartChecking,
        });

        render(<AppRouter />);

        expect(screen.getByText("Loading...")).toBeTruthy();
        expect(mockStartChecking).toHaveBeenCalled();
    });

    test("should show login in case user is not authenticated", () => {
        useAuthStore.mockReturnValue({
            checking: false,
            startChecking: mockStartChecking,
        });

        const { container } = render(<AppRouter />);

        expect(screen.getByText("Log in to IBM")).toBeTruthy();
        expect(container).toMatchSnapshot();
    });

    test("should show main view in case user is authenticated", () => {
        useAuthStore.mockReturnValue({
            checking: false,
            ...testUserInfo,
            startChecking: mockStartChecking,
        });

        render(<AppRouter />);

        expect(screen.getAllByText("Home")).toBeTruthy();
    });
});
