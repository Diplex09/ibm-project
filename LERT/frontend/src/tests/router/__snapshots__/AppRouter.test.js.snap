// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in AppRouter should show login in case user is not authenticated 1`] = `
<div>
  <main
    class="MuiContainer-root MuiContainer-maxWidthFalse css-2bi4m2-MuiContainer-root"
  >
    <div
      class="MuiBox-root css-1fp3p0t"
    >
      <div
        class="MuiBox-root css-ckhcb9"
      >
        <div
          class="MuiBox-root css-c5ephf"
        >
          <div
            class="MuiBox-root css-j7qwjs"
          >
            <span
              class="MuiTypography-root MuiTypography-body1 css-1c7a8s7-MuiTypography-root"
            >
              Welcome to,
            </span>
            <span
              class="MuiTypography-root MuiTypography-h3 css-162y9ut-MuiTypography-root"
            >
              IBM
               
              <span
                class="MuiTypography-root MuiTypography-h3 css-19qgod0-MuiTypography-root"
              >
                LERT
              </span>
            </span>
            <p
              class="MuiTypography-root MuiTypography-body1 css-1x4iig-MuiTypography-root"
            >
              Start using the Labor Expenses Recovery Tool
            </p>
          </div>
          <div
            class="MuiBox-root css-0"
          >
            <span
              class="MuiTypography-root MuiTypography-body1 css-fpxq78-MuiTypography-root"
            >
              Learn more:
            </span>
            <div
              class="MuiBox-root css-9c7r58"
            >
              <a
                class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-s9q0mz-MuiTypography-root-MuiLink-root"
                href="https://www.ibm.com/"
                target="_blank"
              >
                Terms of use
              </a>
              <a
                class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-s9q0mz-MuiTypography-root-MuiLink-root"
                href="https://www.ibm.com/"
                target="_blank"
              >
                IBM Internal Privacy Statement
              </a>
              <a
                class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-s9q0mz-MuiTypography-root-MuiLink-root"
                href="https://www.ibm.com/"
                target="_blank"
              >
                Feedback
              </a>
              <a
                class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-s9q0mz-MuiTypography-root-MuiLink-root"
                href="https://www.ibm.com/"
                target="_blank"
              >
                Business Conduct Guidelines
              </a>
              <a
                class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-s9q0mz-MuiTypography-root-MuiLink-root"
                href="https://www.ibm.com/"
                target="_blank"
              >
                Support
              </a>
            </div>
          </div>
        </div>
        <div
          class="MuiBox-root css-20nurc"
        >
          <h1
            class="MuiTypography-root MuiTypography-h5 css-1v2pkav-MuiTypography-root"
          >
            Log in to IBM
             
            <p
              class="MuiTypography-root MuiTypography-body1 css-wle25o-MuiTypography-root"
            >
              LERT
            </p>
          </h1>
          <form
            class="MuiBox-root css-tz6en3"
            novalidate=""
          >
            <div
              class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
            >
              <label
                class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary Mui-focused Mui-required css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="true"
                for="email"
                id="email-label"
              >
                Email Address
                <span
                  aria-hidden="true"
                  class="MuiInputLabel-asterisk MuiFormLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
                >
                   
                  *
                </span>
              </label>
              <div
                class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth Mui-focused MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-invalid="false"
                  autocomplete="email"
                  class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                  id="email"
                  name="email"
                  required=""
                  type="text"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-1in441m"
                  >
                    <span>
                      Email Address
                       
                      *
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
            <div
              class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
            >
              <label
                class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary Mui-required css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="false"
                for="password"
                id="password-label"
              >
                Password
                <span
                  aria-hidden="true"
                  class="MuiInputLabel-asterisk MuiFormLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
                >
                   
                  *
                </span>
              </label>
              <div
                class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-invalid="false"
                  autocomplete="current-password"
                  class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                  id="password"
                  name="password"
                  required=""
                  type="password"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-1ftyaf0"
                  >
                    <span>
                      Password
                       
                      *
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
            <label
              class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-j204z7-MuiFormControlLabel-root"
            >
              <span
                class="MuiCheckbox-root MuiCheckbox-colorPrimary MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
              >
                <input
                  class="PrivateSwitchBase-input css-1m9pwf3"
                  data-indeterminate="false"
                  type="checkbox"
                  value="remember"
                />
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="CheckBoxOutlineBlankIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                  />
                </svg>
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </span>
              <span
                class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
              >
                Remember me
              </span>
            </label>
            <button
              class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth MuiButtonBase-root  css-3mc3fz-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="submit"
            >
              Log in 
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
                data-testid="ArrowForwardOutlinedIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="m12 4-1.41 1.41L16.17 11H4v2h12.17l-5.58 5.59L12 20l8-8-8-8z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
            <div
              class="MuiGrid-root MuiGrid-container css-11lq3yg-MuiGrid-root"
            >
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true css-1vd824g-MuiGrid-root"
              >
                <a
                  class="MuiTypography-root MuiTypography-body2 MuiLink-root MuiLink-underlineAlways css-wpssva-MuiTypography-root-MuiLink-root"
                  href="#"
                >
                  Forgot password?
                </a>
              </div>
            </div>
          </form>
        </div>
      </div>
    </div>
  </main>
</div>
`;
